### PR TITLE
fixes #3917 - Added Strong Parameters

### PR DESCRIPTION
**Pull request by shlomizadok**

### CODE REVIEW

**Comment by orrabin** [comment]
[test]

File: app/controllers/auth_source_ldaps_controller.rb [hunk 1]
```diff
@@ -23,8 +23,8 @@ def edit
 
   def update
     # remove from hash :account_password if blank?
-    params[:auth_source_ldap].except!(:account_password) if params[:auth_source_ldap][:account_password].blank?
-    if @auth_source_ldap.update_attributes(params[:auth_source_ldap])
+    # params[:auth_source_ldap].except!(:account_password) if params[:auth_source_ldap][:account_password].blank?
```
Comment (dLobatog) [inline app/controllers/auth_source_ldaps_controller.rb]: Why is this commented out?

**Comment by dLobatog** [comment]
@shlomizadok Thanks, this looks like a good add. I haven't fully reviewed the allowed params yet, but before this can go in, I think we should refactor all the `xxxxx_params` methods into the ApplicationController.

``` ruby
private

def architecture_params
  params.require(:architecture).permit(*permitted_architecture_attributes)
end
```

In most cases (thankfully) just defining this in terms of the `controller_name` will suffice. In other cases overriding the ApplicationController method is okay.

**Comment by dLobatog** [comment]
Sorry, I forgot to add new gems, strong_params in this case, need also a PR in [foreman-packaging](https://github.com/theforeman/foreman-packaging) rpm/develop and deb/develop so that our repo can serve the packages for these gems. 

@shlomizadok mind to submit the changes there? [rpm example](https://github.com/theforeman/foreman-packaging/pull/356) [deb example](https://github.com/theforeman/foreman-packaging/pull/360)

**Comment by domcleal** [comment]
Debian packages aren't required, and I'm happy to do the RPM if it's outside your comfort zone.

File: test/functional/architectures_controller_test.rb [hunk 2]
```diff
@@ -23,7 +23,7 @@ def test_new_cancel_button_id
 
   def test_create_invalid
     Architecture.any_instance.stubs(:valid?).returns(false)
-    post :create, {}, set_session_user
+    post :create, {:architecture => {:name => nil}}, set_session_user
```
Comment (dLobatog) [inline test/functional/architectures_controller_test.rb:26]: Why? What happens if you send an empty `{}`? Most likely we want to retain this behavior.

File: test/functional/config_templates_controller_test.rb [hunk 3]
```diff
@@ -49,13 +49,13 @@ class ConfigTemplatesControllerTest < ActionController::TestCase
 
   test "update invalid" do
     ConfigTemplate.any_instance.stubs(:valid?).returns(false)
-    put :update, {:id => config_templates(:pxekickstart).to_param}, set_session_user
+    put :update, {:id => config_templates(:pxekickstart).to_param, :config_template => { :name => config_templates(:pxekickstart).name } }, set_session_user
```
Comment (dLobatog) [inline test/functional/config_templates_controller_test.rb:52]: A key `:config_template` shouldn't be needed for updates, the `:id` is enough.

File: test/functional/smart_proxies_controller_test.rb [hunk 4]
```diff
@@ -31,13 +31,14 @@ def test_edit
...
     SmartProxy.any_instance.stubs(:valid?).returns(false)
-    put :update, {:id => SmartProxy.first}, set_session_user
+    put :update, {:id => SmartProxy.first.to_param, :smart_proxy => {:url => nil}}, set_session_user
     assert_template 'edit'
   end
 
   def test_update_valid
     SmartProxy.any_instance.stubs(:valid?).returns(true)
-    put :update, {:id => SmartProxy.first}, set_session_user
+    put :update, {:id => SmartProxy.first,:smart_proxy => {:url => "http://elsewhere.com:8443"}}, set_session_user
```
Comment (dLobatog) [inline test/functional/smart_proxies_controller_test.rb:40]: Updating with an empty payload should work well too if I remember the RFC correctly.

**Comment by shlomizadok** [comment]
@domcleal -- would love to try to create the RPM (will probably ping you for some help...). But first will do some changes to this commit.

File: app/controllers/interfaces_controller.rb [hunk 5]
```diff
@@ -3,7 +3,7 @@ class InterfacesController < ApplicationController
   # params structure is
   #   {"host"=>
   #     {"interfaces_attributes"=>
-  #       {"new_1405068143746"=>
+  #       {"1405068143746"=>
```
Comment (shlomizadok) [inline app/controllers/interfaces_controller.rb:6]: Changed the JS which creates this attributes. changed the example.

File: test/functional/smart_proxies_controller_test.rb (hunk 4 above)
Comment (shlomizadok) [inline test/functional/smart_proxies_controller_test.rb:40]: `param.require(:param_name)` will fail if `:param_name` is empty.
see: http://api.rubyonrails.org/classes/ActionController/Parameters.html#method-i-require

``` ruby
ActionController::Parameters.new(person: {}).require(:person)
# => ActionController::ParameterMissing: param not found: person
```

File: test/functional/architectures_controller_test.rb (hunk 2 above)
Comment (shlomizadok) [inline test/functional/architectures_controller_test.rb:26]: `param.require(:param_name)` will fail if `:param_name` is empty.
see: http://api.rubyonrails.org/classes/ActionController/Parameters.html#method-i-require

``` ruby
ActionController::Parameters.new(person: {}).require(:person)
# => ActionController::ParameterMissing: param not found: person
```

File: test/functional/config_templates_controller_test.rb (hunk 3 above)
Comment (shlomizadok) [inline test/functional/config_templates_controller_test.rb:52]: `param.require(:param_name)` will fail if `:param_name` is empty.
see: http://api.rubyonrails.org/classes/ActionController/Parameters.html#method-i-require

``` ruby
ActionController::Parameters.new(person: {}).require(:person)
# => ActionController::ParameterMissing: param not found: person
```

File: app/assets/javascripts/application.js [hunk 6]
```diff
@@ -101,6 +101,17 @@ function onContentLoad(){
   });
 
   multiSelectOnLoad();
+
+  // Removes all the templates from forms as they can't be cleared by strong parameters.
+  $('form').on('submit', function(e) {
+    $('div.lookup_keys_fields_template').remove();
+    $('div.lookup_values_fields_template').remove();
```
Comment (ohadlevy) [inline app/assets/javascripts/application.js]: cant we simply add it in the lookup_key js (AFAIR thats where its added in the first place)

File: app/assets/javascripts/lookup_keys.js [hunk 7]
```diff
@@ -79,7 +79,7 @@ function add_child_node(item) {
     // Make a unique ID for the new child
     var regexp  = new RegExp('new_' + assoc, 'g');
     var new_id  = new Date().getTime();
-    content     = content.replace(regexp, "new_" + new_id);
+    content     = content.replace(regexp, new_id);
```
Comment (ohadlevy) [inline app/assets/javascripts/lookup_keys.js]: is this still needed?

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -387,4 +388,9 @@ def two_pane?
   def handle_unverified_request
     raise ::Foreman::Exception.new(N_("Invalid authenticity token"))
   end
+
+  def foreman_params
+    permitted_params = "permitted_#{controller_name.singularize}_attributes"
+    params.require(controller_name.singularize.to_sym).permit(*send(permitted_params))
```
Comment (ohadlevy) [inline app/controllers/application_controller.rb:395]: maybe add a check if it response to permitter_params?

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -18,6 +18,7 @@ def new
 
   def create
     if @user.save
+      update_sub_hostgroups_owners
```
Comment (ohadlevy) [inline app/controllers/users_controller.rb]: where does this comes from?

File: app/helpers/strong_parameters_helper.rb [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+  def permitted_usergroup_attributes
+    permitted_attributes.usergroup_attributes +
+        [:external_usergroups_attributes => [:name, :auth_source_id] ]
+  end
+
+  def vm_attributes
+    [
+        ## case ComputeResource.type ??
+        # Libvirt
```
Comment (ohadlevy) [inline app/helpers/strong_parameters_helper.rb:76]: how would a plugin extend this?

File: app/views/hostgroups/_form.html.erb [hunk 11]
```diff
@@ -71,6 +71,5 @@
     <%= render "taxonomies/loc_org_tabs", :f => f, :obj => @hostgroup %>
   </div>
 
-  <%= f.hidden_field :id %>
```
Comment (ohadlevy) [inline app/views/hostgroups/_form.html.erb:76]: this was originally added by @eLobato but I assume its not really required?

File: config/environments/development.rb [hunk 12]
```diff
@@ -34,5 +34,9 @@
   # Log the query plan for queries taking more than this (works
   # with SQLite, MySQL, and PostgreSQL)
   config.active_record.auto_explain_threshold_in_seconds = 0.5
+
+  # For development of Strong Params, this has been put in development.rb
+  # Need to consider putting it env-wise or initializer
+  config.action_controller.action_on_unpermitted_parameters = :raise
```
Comment (ohadlevy) [inline config/environments/development.rb:48]: what is the default for prod?

**Comment by ohadlevy** [comment]
@shlomizadok first cut looks good, I'm mostly missing tests and plugin extension ability.
also, can you please create a test that fails if the user forget to add a new mapping if a new model/controller are added (see similar implementation for permissions).

File: config/environments/development.rb (hunk 12 above)
Comment (shlomizadok) [inline config/environments/development.rb:48]: If not set up, in development & test unpermitted parameters will be logged. On production, unpermitted parameters will be filtered out.
See [strong parameters gem](https://github.com/rails/strong_parameters#handling-of-unpermitted-keys).

File: app/views/hostgroups/_form.html.erb (hunk 11 above)
Comment (shlomizadok) [inline app/views/hostgroups/_form.html.erb:76]: Tested. Not required.

File: app/assets/javascripts/lookup_keys.js (hunk 7 above)
Comment (shlomizadok) [inline app/assets/javascripts/lookup_keys.js]: Yes it is. This creates the new lookup_key fields. I have removed the word "new_" as when clearing parameter for nested attributes expect to get numbered hash. e.g.,

``` ruby
{ 0 => {:important => :attribute}, 1=> {:_destory => true}}
```

This code creates a new field, with number, so strong params can handle it.

**Comment by shlomizadok** [comment]
@ohadlevy added some tests. Some show how to extend strong parameters via plugin extension.
Currently fails on Katello.

**Comment by shlomizadok** [comment]
Rebased

**Comment by unorthodoxgeek** [comment]
@eLobato @ohadlevy - what's the status on this? I would really want to have this thing merged, as it's a blocker for rails4

**Comment by ohadlevy** [comment]
@unorthodoxgeek i think its pending review, would you mind going over it? ty

File: app/helpers/strong_parameters_helper.rb (hunk 10 above)
Comment (unorthodoxgeek) [inline app/helpers/strong_parameters_helper.rb:76]: would this work?

``` ruby
module StrongParametersHelper
  alias_method :vm_attributes_without_clound_provider, :vm_attributes
  def vm_attributes
    vm_attributes_without_clound_provider + attributes
  end
end
```

File: test/functional/strong_parameters_test.rb [hunk 13]
```diff
@@ -0,0 +1,43 @@
+require 'test_helper'
```
Comment (unorthodoxgeek) [inline test/functional/strong_parameters_test.rb:1]: I really dislike this. you added a route which is pointless except in tests, and created a controller, which has no use except in tests.
the strong_parameters gem is well tested, we don't need to retest it. you can test that gem's integration with real controllers, that would make more sense.

**Comment by unorthodoxgeek** [comment]
my main concern is that this is basically going to break all the plugins. we're okay with that?

File: test/functional/strong_parameters_test.rb (hunk 13 above)
Comment (shlomizadok) [inline test/functional/strong_parameters_test.rb:1]: The purpose of this test is to test if external (e.g plugins) can use foreman_params and extend to their own attributes. This does not test strong_parameters.

Comment (unorthodoxgeek) [inline test/functional/strong_parameters_test.rb:1]: ok :+1:

**Comment by shlomizadok** [comment]
This will not break all plugins (only some of them).
The plugins that use deface to add attributes to existing web ui controllers may break.
Other plugins who have their own controllers will not break, but it will be a good practice to refactor them to use strong parameters.

**Comment by dLobatog** [comment]
@shlomizadok Closing as this is pretty much already in the Rails 4 branch - so this code won't be lost but merged at somepoint. Reopen if you want to keep discussing. Thanks!

**Pull request closed without merging**